caikit/interfaces/ts/data_model/backends/_spark_backends.py

```diff
@@ -0,0 +1,208 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Core data model backends backed by pyspark.sql.DataFrame.
+
+This module is not intended for direct importing. It's used
+by the caikit ts datamodel. Directly importing this module
+will force a hard spark dependency which we do not want
+to do.
+"""
+
+# Standard
+from typing import Any, Iterable, Optional, Tuple, Type, Union
+
+# Third Party
+import pandas as pd
+
+# this import is ok because this module is NOT proactively imported
+import pyspark
+
+# First Party
+import alog
+
+# Local
+from .....core.data_model import ProducerId
+from .....core.exceptions import error_handler
+from .._single_timeseries import SingleTimeSeries
+from .base import MultiTimeSeriesBackendBase, TimeSeriesBackendBase
+from .dfcache import EnsureCached
+from .pandas_backends import PandasMultiTimeSeriesBackend, PandasTimeSeriesBackend
+from .util import mock_pd_groupby
+
+log = alog.use_channel("SPBCK")
+error = error_handler.get(log)
+
+
+class SparkMultiTimeSeriesBackend(MultiTimeSeriesBackendBase):
+    def __init__(
+        self,
+        data_frame: pyspark.sql.DataFrame,
+        key_column: Union[Iterable[str], str],
+        timestamp_column: str = None,
+        value_columns: Optional[Iterable[str]] = None,
+        ids: Optional[Union[Iterable[int], Iterable[str]]] = None,
+        producer_id: Optional[Union[Tuple[str, str], ProducerId]] = None,
+    ):
+        error.type_check("<COR77829913F>", pyspark.sql.DataFrame, data_frame=data_frame)
+
+        # for param validation
+        _ = PandasMultiTimeSeriesBackend(
+            data_frame=pd.DataFrame(columns=data_frame.columns),
+            key_column=key_column,
+            timestamp_column=timestamp_column,
+            value_columns=value_columns,
+            ids=ids,
+            producer_id=producer_id,
+        )
+
+        self._pyspark_df: pyspark.sql.DataFrame = data_frame
+        # for tapping into pandas api call when needed
+        self._pyspark_pandas_df = self._pyspark_df.pandas_api()
+        self._key_column = key_column
+        self._timestamp_column = timestamp_column
+        # pylint: disable=duplicate-code
+        self._value_columns = self._value_columns = value_columns or [
+            col
+            for col in data_frame.columns
+            if col != timestamp_column and col not in key_column
+        ]
+        self._ids = [] if ids is None else ids
+        self._producer_id = (
+            producer_id
+            if isinstance(producer_id, ProducerId)
+            else (ProducerId(*producer_id) if producer_id is not None else None)
+        )
+
+    def get_attribute(
+        self, data_model_class: Type["MultiTimeSeries"], name: str
+    ) -> Any:
+        # pylint: disable=duplicate-code
+        if isinstance(self._key_column, str):
```

Comment:
Moved this for `pandas_backends.py` and then used the instance of `PandasMultiTimeSeriesBackend` in `_spark_backends.py` to avoid the code duplication.